ext/ddtrace_profiling_native_extension/heap_recorder.c

```diff
@@ -0,0 +1,489 @@
+#include "heap_recorder.h"
+#include <pthread.h>
+#include "ruby/st.h"
+#include "ruby/util.h"
+#include "ruby_helpers.h"
+#include <errno.h>
+
+#define MAX_FRAMES_LIMIT 10000
+#define MAX_QUEUE_LIMIT 10000
+
+// Initial seed for hash functions
+#define FNV1_32A_INIT 0x811c9dc5
+
+typedef struct {
+  char *name;
+  char *filename;
+  int64_t line;
+} heap_frame;
+static int heap_frame_cmp(heap_frame*, heap_frame*);
+static st_index_t heap_frame_hash(heap_frame*, st_index_t seed);
+
+typedef struct {
+  heap_frame *frames;
+  uint64_t frames_len;
+} heap_stack;
+static heap_stack* heap_stack_init(ddog_prof_Slice_Location);
+static void heap_stack_free(heap_stack*);
+void heap_stack_debug(const heap_stack*);
+static int heap_stack_cmp(heap_stack*, heap_stack*);
+static st_index_t heap_stack_hash(heap_stack*, st_index_t);
+static int heap_stack_cmp_st(st_data_t, st_data_t);
+static st_index_t heap_stack_hash_st(st_data_t);
+static const struct st_hash_type st_hash_type_heap_stack = {
+    heap_stack_cmp_st,
+    heap_stack_hash_st,
+};
+
+typedef struct {
+  // How many objects are currently tracked by the heap recorder for this heap record.
+  uint64_t num_tracked_objects;
+  // Estimate for how many objects are currently in use for this heap record.
+  // (this is basically a sum of the weights of the num_tracked_objects)
+  uint64_t inuse_objects;
+  const heap_stack *stack;
+} heap_record;
+static heap_record* heap_record_init(const heap_stack*);
+static void heap_record_free(heap_record*);
+
+typedef struct {
+  VALUE obj;
+  unsigned int weight;
+  heap_record *heap_record;
+} object_record;
+static object_record* object_record_init(VALUE, unsigned int, heap_record*);
+static void object_record_free(object_record*);
+
+typedef struct {
+  VALUE obj;
+  unsigned int weight;
+  ddog_CharSlice *class_name;
```

Comment:
We're not using the class name yet, but do note this pointer is pointing at something that's stack-allocated, so it's only valid while taking the sample (aka it'll need to be copied if we want to keep it for later).